src/toil/job.py

```diff
@@ -157,11 +159,65 @@ def _parseResource(name, value):
                             % (name, type(value)))
 
 
-class Job(ResourceRequirementMixin):
+class JobLikeObject(ResourceRequirementMixin):
+    def __init__(self, name, job=None, **requirements):
+        super(JobLikeObject, self).__init__(**requirements)
+        self.name = name
+        self.job = job or self.__class__.__name__
+
+    def __str__(self):
+        printedName = '\''+self.job + '\''
+        if self.name:
+            printedName += ' '+self.name
+        return str(printedName)
+
+class JobNode(JobLikeObject):
+    """
+    This object bridges the job graph, job, and batchsystem classes
+    """
+    def __init__(self, memory, cores, disk, preemptable, job, name, jobStoreID,
+                 command, predecessorNumber=1):
+        super(JobNode, self).__init__(memory=memory, cores=cores, disk=disk,
+                                      preemptable=preemptable, name=name, job=job)
+        self.jobStoreID = jobStoreID
+        self.predecessorNumber = predecessorNumber
+        self.command = command
+
+    def __hash__(self):
+        return hash(self.jobStoreID)
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __repr__(self):
+        return '%s( **%r )' % (self.__class__.__name__, self.__dict__)
+
+    @classmethod
+    def fromJobGraph(cls, jobGraph):
+        return cls(jobStoreID=jobGraph.jobStoreID, memory=jobGraph.memory,
+                   cores=jobGraph.cores, disk=jobGraph.disk,
+                   preemptable=jobGraph.preemptable,
+                   command=jobGraph.command,
+                   job=jobGraph.job,
+                   name=jobGraph.name,
+                   predecessorNumber=jobGraph.predecessorNumber)
+
+    @classmethod
+    def fromJob(cls, job, command, predecessorNumber):
+        return cls(jobStoreID=None, memory=job.memory,
+                   cores=job.cores, disk=job.disk,
+                   preemptable=job.preemptable,
+                   command=command,
+                   job=job.job,
+                   name=job.name,
+                   predecessorNumber=predecessorNumber)
+
+
+class Job(JobLikeObject):
     """
     Class represents a unit of work in toil.
     """
-    def __init__(self, memory=None, cores=None, disk=None, preemptable=None, checkpoint=False):
+    def __init__(self, memory=None, cores=None, disk=None, preemptable=None, name='', checkpoint=False):
```

Comment:
Again, empty string vs. None ugliness. Basically, don't treat empty string like None, but rather like any other string and make sure that `__str__` and `__repr__` return values that make it obvious that an empty string was used for either `name` or `job`.